File: y2019/haskell/Day1.hs
import System.IO
import Data.Char
import Text.Printf

main = do
    file <- readFile "day1/day1.txt"
    let masses :: [Integer] = map read (lines file)

    let needed m = max (m `div` 3 - 2) 0;
    let fuelSum1 = sum (map needed masses);

    printf "amt of fuel needed for modules (p1): %d\n" fuelSum1

    let totalNeeded m = if m == 0 then 0
                        else (let f = needed m in f + totalNeeded f)
    let fuelSum2 = sum (map totalNeeded masses);
    printf "amt of fuel needed for modules (p2): %d\n" fuelSum2

File: y2019/go/day1.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strconv"
)

func fuelNeeded(mod int) int {
	return max(mod/3-2, 0)
}

func day1() {
	file, err := os.Open("../input/day1.txt")
	if err != nil {
		log.Fatal(err)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	var modules []int
	for scanner.Scan() {
		m, _ := strconv.Atoi(scanner.Text())
		modules = append(modules, m)
	}

	p1Fuel := 0
	p2Fuel := 0
	for _, m := range modules {
		p1Fuel += fuelNeeded(m)
		for m > 0 {
			p2Fuel += fuelNeeded(m)
			m = fuelNeeded(m)
		}
	}

	fmt.Printf("amt of fuel for p1: %v\n", p1Fuel)
	fmt.Printf("amt of fuel for p2: %v\n", p2Fuel)
}
